feature(backend): refactor user command with support of username (#3093)
tracim/tracim: backend/doc/setting.md

The last one will check the internal user database first. Then, if the auth fails, it will also try to authenticate the user using LDAP.

> The authentication can be done with a login, a login can be either email or username of user. In case there is

Reviewer: ```suggestion
The authentication can be done with a login. A login can be either the user's email address or the username. In case there is
```

> no clear distinction in the authentication method used, login with "@" will be considered as email and login without as username.

Reviewer: Is this possible that there is an ambiguity though? `@` is forbidden in the username. I would drop this sentence entirely.

Author: The Api auth method does make distinction between email and username field, other auth method that require login use "login" which will check either email or username depending if "@" is provided.

Reviewer: Ah, ok. I didn't understand this sentence in this sense. I'll add a suggestion

> :warning: If you use LDAP or Remote Auth method, the automatic creation of username user (without "@") can failed if
> you have set "email.required" to True, which means every user should have an email set, To resolve this case, you can either
> set email.required to False, or create user with both username and email and then authenticate using LDAP/Remote Auth.

Reviewer: ```suggestion
:warning: If you use LDAP or the Remote Auth method, the automatic creation of a username-only user (without "@") can fail if:
`email.required=True`, which means every user should have an email address set. To solve this case, either:
- set `email.required` to `False`, or
- create the user with both an username and an email address and then authenticate using the LDAP/Remote Auth.
```

The last one will check the internal user database first. Then, if the auth fails, it will also try to authenticate the user using LDAP.

> The authentication can be done with a login. A login can be either the user's email address or the username. In case there is
> no clear distinction in the authentication method used, login with "@" will be considered as email and login without as username.

Reviewer: ```suggestion
The authentication is done with a login, which is either the user's email address or the username. For authentication methods which don't provide a way to distinguish between the two, a login with "@" will be considered as an email address and a login without as username.
```

@inkhey what about usernames typed as `@someone` *?

Author: This is not supported in backend now
In current implementation, it's the frontend that does the jobs to deal with leading "@". We can handle this in backend but that either mean handled it deeply in tracimlib which is inconsistent with the fact that frontend does the job of removing it, or this mean specific code for each auth method (which is not nice).

I don't think it's a required feature to handle this for technical auth mecanism.

Author: By the way, there was a discussion about this here: https://github.com/tracim/tracim/issues/3061

It relies on 2 HTTP headers:

- `Tracim-Api-Key` : Tracim api key, as marked in config in `api.key`

> - `Tracim-Api-Login` : User email/username login, in order to act as the user given

Reviewer: ```suggestion
- `Tracim-Api-Login` : User's login (either an email or a username), in order to act as the given user
```

If `api.key` is empty, the API key authentication will be disabled.

### Remote Auth Authentification (eg apache authentication)

It is possible to connect to Tracim using remote authentification (e.g. the Apache authentication method).

> The idea is that the webserver authenticates user and then pass the login of the authenticated user trhough uWSGI environment variables or an HTTP header.

Reviewer: ```suggestion
The idea is that the webserver authenticates the user and then pass the login of the authenticated user through uWSGI environment variables or an HTTP header.
```
